packages/core/src/events/embeddings/base.ts

```diff
@@ -1,36 +1,213 @@
+import { BaseCache } from '../../cache/base.js';
+import { MemoryCache } from '../../cache/index.js';
+import { CallableConfig } from '../../record/callable.js';
 import {
-    AsyncCallError,
-    AsyncCaller,
-    type AsyncCallerParams,
-  } from '../../utils/asyncCaller.js';
-
-export type EmbeddingsParams = AsyncCallerParams;
-
-export abstract class Embeddings {
-    /**
-     * The async caller should be used by subclasses to make any async calls,
-     * which will thus benefit from the concurrency and retry logic.
-     */
-    caller: AsyncCaller;
-  
-    constructor(params: EmbeddingsParams) {
-      this.caller = new AsyncCaller(params ?? {});
+  AsyncCaller,
+  type AsyncCallerParams,
+} from '../../utils/asyncCaller.js';
+import { BaseEvent, BaseEventParams } from '../base.js';
+import { EmbedResult } from '../output/provide/embedresult.js';
+
+export interface BaseEmbeddingsCallOptions extends BaseEventParams {
+  /**
+   * Abort signal for the call.
+   * If provided, the call will be aborted when the signal is aborted.
+   */
+  signal?: AbortSignal;
+}
+
+export interface BaseEmbeddingsParams
+  extends AsyncCallerParams,
+    BaseEventParams {
+  cache?: BaseCache<number[]> | boolean;
+}
+
+export abstract class BaseEmbeddings<
+    CallOptions extends BaseEmbeddingsCallOptions = BaseEmbeddingsCallOptions,
+  >
+  extends BaseEvent<string, EmbedResult, CallOptions>
+  implements BaseEmbeddingsParams
```

Comment:
Is there a particular reason why BaseEmbedding implements BaseEmbeddingsParams?

Reply:
You can think `BaseEmbeddingsParams` as the interface to the `BaseEmbedding` class